### PR TITLE
Add conv for torchx

### PR DESCRIPTION
**Pull request by howard0su**

### CODE REVIEW

File: torchx/c_src/torchx.cpp [hunk 1]
```diff
@@ -786,6 +786,32 @@ NIF(solve)
...
+  c10::optional<at::Tensor> bias_tensor;
+
+  std::vector<int64_t> output_padding;
+  // aten::_convolution(Tensor input, Tensor weight, Tensor? bias,
+  //      int[] stride, int[] padding, int[] dilation, bool transposed,
+  //      int[] output_padding, int groups, bool benchmark, bool deterministic, bool cudnn_enabled, bool allow_tf32) -> Tensor
+  TENSOR(at::_convolution(*tensorA, *tensorB, bias_tensor,
+    stride, padding, dilation, transposed, output_padding, groups,
+    true,   // benchmark
+    true,   // deterministic
```
Comment (seanmor5) [inline torchx/c_src/torchx.cpp:809]: Just a note that for performance this should probably be set to false, but we should later consider or see how we can add a deterministic API to Nx compilers/backends to guarantee consistent results in different algorithms

Comment (howard0su) [reply torchx/c_src/torchx.cpp:809]: ok

File: torchx/lib/torchx/backend.ex [hunk 2]
```diff
@@ -848,6 +848,30 @@ defmodule Torchx.Backend do
...
+    # output_permutation = opts[:output_permutation]
+
+    if padding == :valid do
+      padding = List.duplicate(0, length(t.shape))
+    end
+
+    t = from_nx(t)
+    k = from_nx(k)
+
+    to_nx(Torchx.conv(t, k, strides, padding, input_dilation, false, feature_groups), out)
```
Comment (polvalente) [inline torchx/lib/torchx/backend.ex:872]:
```suggestion
    t
    |> Torchx.conv(k, strides, padding, input_dilation, false, feature_groups)
    |> to_nx(out)
```

File: torchx/lib/torchx/backend.ex [hunk 3]
```diff
@@ -870,6 +870,36 @@ defmodule Torchx.Backend do
...
+    # Not supported yet
+    # input_dilation = opts[:input_dilation]
+    # batch_groups = opts[:batch_group_size]
+    # input_permutation = opts[:input_permutation]
+    # kernel_permutation = opts[:kernel_permutation]
+    # output_permutation = opts[:output_permutation]
+
+    pad_config =
+      padding
+      |> Enum.map(fn {a, b} -> [a, b] end)
```
Comment (howard0su) [inline torchx/lib/torchx/backend.ex:889]: shall I use Tuple.to_list here? which one is better?

Comment (josevalim) [reply torchx/lib/torchx/backend.ex:889]: Both are fine. Given this case you always have two elements, I prefer the current version!

File: torchx/lib/torchx/backend.ex [hunk 4]
```diff
@@ -870,6 +870,36 @@ defmodule Torchx.Backend do
...
+    strides = opts[:strides]
+    kernel_dilation = opts[:kernel_dilation]
+    feature_groups = opts[:feature_group_size]
+
+    # Not supported yet
+    # input_dilation = opts[:input_dilation]
+    # batch_groups = opts[:batch_group_size]
+    # input_permutation = opts[:input_permutation]
+    # kernel_permutation = opts[:kernel_permutation]
+    # output_permutation = opts[:output_permutation]
```
Comment (josevalim) [inline torchx/lib/torchx/backend.ex:885]: Should we raise if any of those options are set?

File: nx/lib/nx.ex [hunk 5]
```diff
@@ -7823,6 +7823,71 @@ defmodule Nx do
...
+              [25.0],
+              [20.0]
+            ],
+            [
+              [35.0],
+              [30.0]
+            ]
+          ]
+        ]
+      >
```
Comment (josevalim) [inline nx/lib/nx.ex:7889]: Let's please move this test to torchx/test/nx_test.exs. My concern about having it here is that it is too long and therefore not very useful as documentation. :(

File: torchx/lib/torchx/backend.ex [hunk 6]
```diff
@@ -870,6 +870,57 @@ defmodule Torchx.Backend do
...
 
+  @impl true
+  def conv(%T{type: type} = out, t, k, opts) do
+    unsupported_option!(opts, :batch_group_size, 1)
+
+    input_dilation = opts[:input_dilation]
+
+    Enum.each(input_dilation, fn a ->
+      if a != 1 do
+        raise "input_dilation other than 1 is not supported"
```
Comment (josevalim) [inline torchx/lib/torchx/backend.ex:881]:
```suggestion
        raise ArgumentError,
              "input_dilation other than 1 is not supported, got: #{inspect(input_dilation)}"
```

**Review by josevalim — COMMENTED** [review]
This is looking great. Just two minor comments and we can ship it!!

File: nx/lib/nx.ex (hunk 5 above)
Comment (howard0su) [reply nx/lib/nx.ex:7889]: This should be a generic tests not specific to torchx.

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**